File: doubly-linked-list.js
class Node {
    constructor(previous, data, next) {
        this.previous = previous
        this.data = data;
        this.next = next
    }
}
class DoubleLinkedList {
    constructor() {
        this.head = null;
        this.tail = null
        this.length = 0;
    }
    unshift(data) {
        if (!this.head) {
            const newNode = new Node(null, data, this.head)
            this.head = newNode
            this.tail = this.head
            this.length++
        }
        else {
            const formerHeadNode = this.head
            const newNode = new Node(null, data, this.head)
            this.head = newNode
            formerHeadNode.previous = newNode
            this.length++
        }
    }

    // getFirst
    getFirst() {
        return this.head
    }
    // getLast
    getLast() {
        return this.tail
    }
    // clear
    clear() {
        this.head = null
        this.tail = null
        this.length = 0
        return this.head
    }
    // shift
    shift() {
        if (!this.head) {
            return null
        }
        if (this.length == 1) {
            const formerHead = this.head;
            this.head = null
            this.tail = null
            this.length--
            return formerHead
        }
        const formerHead = this.head;
        this.head = this.head.next
        this.head.previous = null
        this.length--
        return formerHead
    }
    // pop
    pop() {
        if (!this.head.next) {
            return this.shift()
        }
        if (!this.head) return null
        this.length--
        const lastNode = this.tail
        this.tail.previous.next = null
        this.tail = this.tail.previous
        return lastNode
    }
    // push
    push(data) {
        if (!this.head) {
            return this.unshift(data)
        }
        let lastNode = this.tail
        const newNode = new Node(lastNode, data, null)
        lastNode.next = newNode
        this.tail = newNode
        this.length++
    }


    // get
    get(index) {
        if (index < 0 || index > this.length - 1) return false
        let count = 0;
        let currentNode = this.head
        while (count < index) {
            currentNode = currentNode.next
            count++
        }
        return currentNode
    }

    // set
    set(data, index) {
        if (index < 0 || index > this.length - 1) return false
        const nodeToReplace = this.get(index)
        nodeToReplace.data = data
        return true
    }

    size() {
        return this.length
    }

    // remove
    remove(index) {
        if (index === 0) {
            this.shift()
            return true

        }
        if (index < 0 || index > this.length - 1) return false
        let prevNode = this.get(index)
        prevNode.previous.next = prevNode.next
        prevNode = null
        return true
    }


    // insert
    insert(data, index) {
        if (index === 0) {
            this.unshift(data)
            return true

        }

        if (index < 0 || index > this.length) return false

        if (index == this.length) {
            this.push(data)
            return true
        }
        else {
            const prevNode = this.get(index - 1)
            const newNode = new Node(prevNode, data, prevNode.next)
            prevNode.next = newNode
            this.length++
            return true
        }



    }
}



const doubleLinkedListSample = new DoubleLinkedList()
// doubleLinkedListSample.push(9)
// doubleLinkedListSample.push(1)
// doubleLinkedListSample.push(2)
// // doubleLinkedListSample.push(0)
// console.log(doubleLinkedListSample)
// console.log(doubleLinkedListSample.pop())
// console.log(doubleLinkedListSample)
// doubleLinkedListSample.insert(10,2)
// console.log(doubleLinkedListSample)
// console.log(doubleLinkedListSample.pop())
// console.log(doubleLinkedListSample.getLast())
// c

module.exports = DoubleLinkedList

File: doubly-linked-list.go
package main

import (
	"log"
	"strconv"
)

type DoublyLinkedList struct {
	size int
	head *Node
	tail *Node
}

type Node struct {
	prev *Node
	data int
	next *Node
}

func (list *DoublyLinkedList) Clear() bool {
	trav := list.head
	if list.IsEmpty() {
		return true
	}
	for trav != nil {
		next := trav.next
		trav.next, trav.prev = nil, nil

		trav = next
	}
	list.head, list.tail, trav = nil, nil, nil
	list.size = 0
	return true
}

func (list *DoublyLinkedList) ListSize() int {
	return list.size
}

func (list *DoublyLinkedList) IsEmpty() bool {
	return list.ListSize() == 0
}

func (list *DoublyLinkedList) Add(data int) bool {
	return list.AddLast(data)
}

// check if index is valid
func (list *DoublyLinkedList) IndexValid(index int) bool {
	if index < 0 || index > list.size-1 {
		return false
	}
	return true
}

// add an element to the head of a doubly linked list 0(1)
func (list *DoublyLinkedList) AddFirst(data int) int {
	// the linked list is empty
	if list.IsEmpty() {
		newNode := &Node{nil, data, nil}
		list.head, list.tail = newNode, newNode
	} else {
		list.head.prev = &Node{nil, data, list.head}
		list.head = list.head.prev
	}
	list.size = list.size + 1
	return data
}

// add an element to the tail of a doubly linked list 0(1)
func (list *DoublyLinkedList) AddLast(data int) bool {
	// the linked list is empty
	if list.IsEmpty() {
		newNode := &Node{nil, data, nil}
		list.head, list.tail = newNode, newNode
	} else {
		list.tail.next = &Node{list.tail, data, nil}
		list.tail = list.tail.next
	}
	list.size = list.size + 1
	return true
}

// check the value of the first node if it exists, 0(1)
func (list *DoublyLinkedList) PeekFirst() int {
	// the linked list is empty
	if list.IsEmpty() {
		log.Fatal("Empty List")
		return 0
	} else {
		return list.head.data
	}
}

//check the value of the last node if it exists, 0(1)
func (list *DoublyLinkedList) PeekLast() int {
	// the linked list is empty
	if list.IsEmpty() {
		log.Fatal("Empty List")
		return 0
	} else {
		return list.tail.data
	}
}



//  remove the first value at the head of a doubly-linked list, 0(1)
func (list *DoublyLinkedList) RemoveFirst() int {
	// the linked list is empty
	if list.IsEmpty() {
		log.Fatal("Empty List")
	}
	data := list.head.data
	list.head = list.head.next
	list.head.prev = nil

	if list.IsEmpty() {
		list.tail = nil
	}
	list.size = list.size - 1
	return data
}

//  remove the last value at the tail of a doubly-linked list, 0(1)
func (list *DoublyLinkedList) RemoveLast() int {
	// the linked list is empty
	if list.IsEmpty() {
		log.Fatal("Empty List")
	}
	data := list.tail.data
	list.tail = list.tail.prev
	list.tail.next = nil

	if list.IsEmpty() {
		list.head = nil
	}
	list.size = list.size - 1
	return data
}

//  remove a particular node from the linked list 0(1)
func (list *DoublyLinkedList) Remove(node *Node) int {
	// check if the node is somewhere at the tail or at the head
	if node.prev == nil {
		return list.RemoveFirst()
	}
	if node.next == nil {
		return list.RemoveLast()
	}
	//make pointers of adjacent nodes skip over "node"
	node.next.prev = node.prev
	node.prev.next = node.next
	// reduce size by one
	list.size = list.size - 1
	data := node.data
	//memory clean up
	node.prev, node.next = nil, nil
	return data
}

//  remove a node at a particular index 0(n)
func (list *DoublyLinkedList) RemoveAt(index int) int {
	// check the index range and make sure its valid
	if !list.IndexValid(index) {
		log.Fatal("Illegal Argument: Index is not within range")
	}
	// this will point to the address of the node we want to remove
	var node_to_remove *Node
	// if index is less than half the size of the link list, search from the front
	if index <= (list.size-1)/2 {
		trav := list.head
		for i := 0; i < index; i++ {
			trav = trav.next
		}
		node_to_remove = trav

	} else {
		// if index is greater than half the size of the link list, search from the back
		trav := list.tail
		for i := list.size - 1; i > index; i-- {
			trav = trav.prev
		}
		node_to_remove = trav

	}

	return list.Remove(node_to_remove)
}


//check the value at a particular node, 0(n)
func (list *DoublyLinkedList) Get(index int) int {
	// check the index range and make sure its valid
	if !list.IndexValid(index) {
		log.Fatal("Illegal Argument: Index is not within range")
	}
	// this will point to the address of the node we want to peek
	var node *Node
	// if index is less than half the size of the link list then search from the front
	if index <= (list.size-1)/2 {
		trav := list.head
		for i := 0; i < index; i++ {
			trav = trav.next
		}
		node = trav

	} else {
		// if index is greater than half the size of the link list, then search from the back

		trav := list.tail
		for i := list.size - 1; i > index; i-- {
			trav = trav.prev
		}
		node = trav
	}

	return node.data
}

// get index of a value in the linked list, 0(n)
func (list *DoublyLinkedList) IndexOf(val int) int {
	// the linked list is empty
	if list.IsEmpty() {
		log.Fatal("Empty List")
	}
	trav := list.head
	index := 0
	for trav != nil {
		if trav.data == val {
			return index
		}
		trav = trav.next
		index = index + 1
	}

	return -1
}

// check if a linked list contains a value, 0(n)
func (list *DoublyLinkedList) Contains(val int) bool {
	// the linked list is empty
	if list.IsEmpty() {
		return false
	}
	trav := list.head
	for trav != nil {
		if trav.data == val {
			return true
		}
		trav = trav.next
	}

	return false
}

// Print the linked list as a string
func (list *DoublyLinkedList) ToString() string {
	list_to_string := ""
	// the linked list is empty
	if list.IsEmpty() {
		return ""
	}
	trav := list.head
	for trav != nil {
		if trav.next != nil {
			list_to_string += strconv.Itoa(trav.data) + "->"
		} else {
			list_to_string += strconv.Itoa(trav.data)
		}

		trav = trav.next
	}
	return list_to_string
}

